File: python/sliding_window/substring_with_concatenation.py
from collections import Counter


class Solution:
    def findSubString(self, s: str, words: list[str]) -> list[str]:
        if not s or not words:
            return []

        word_len = len(words[0])
        word_count = len(words)
        total_len = word_len * word_count
        word_map = Counter(words)

        result = []

        for i in range(word_len):
            left = i
            right = i

            curr_map = Counter()

            while right + word_len <= len(s):
                word = s[right : right + word_len]

                right += word_len

                if word in word_map:
                    curr_map[word] += 1

                    while curr_map[word] > word_map[word]:
                        left_word = s[left : left + word_len]

                        curr_map[left_word] -= 1

                        left += word_len

                    if right - left == total_len:
                        result.append(left)
                else:
                    curr_map.clear()
                    left = right

        return result

File: golang/sliding_window/substring_with_concatenation.go
package main

func findSubstring(s string, words []string) []int {
	if len(s) == 0 || len(words) == 0 {
		return []int{}
	}

	wordLen := len(words[0])
	wordCount := len(words)
	totalLen := wordLen * wordCount

	wordMap := make(map[string]int)

	result := []int{}

	for _, word := range words {
		wordMap[word]++
	}

	for i := 0; i < wordLen; i++ {
		left, right := i, i

		currMap := make(map[string]int)

		for right+wordLen <= len(s) {
			word := s[right : right+wordLen]
			right += wordLen

			if count, found := wordMap[word]; found {
				currMap[word]++

				for currMap[word] > count {
					leftWord := s[left : left+wordLen]

					currMap[leftWord]--
					left += wordLen
				}

				if right-left == totalLen {
					result = append(result, left)
				}
			} else {
				currMap = make(map[string]int)

				left = right
			}
		}
	}

	return result
}
